File: jarvis-desktop/jarvis_desktop/actions/builtins.py
"""The LOCAL action table — the authority on what this machine can do and how
dangerous each of those things is.

The server's ``tier`` field can raise these numbers for a single command;
nothing can lower them. Adding an action is a two-line change here plus a row in
the README, so the table is deliberately in one place where it can be reviewed
as a whole.

The tiers here match the phone's table action for action where the two overlap
(``read_file`` AUTO, ``write_file`` NOTIFY, ``delete_file`` CONFIRM,
``read_clipboard``/``write_clipboard`` NOTIFY, ``http_request`` NOTIFY raising
to CONFIRM for non-GET, shell CONFIRM). That is the point: a user who learns
what Jarvis will and will not do without asking should not have to learn it
twice.
"""

from __future__ import annotations

from ..audit import AuditLog
from ..config import Config
from ..consent import ConsentGateway
from ..policy import ActionTier, PolicyProvider
from .apps import FocusWindow, LaunchApp, ListWindows, OpenUrl
from .base import Action, ActionContext
from .clipboard import ReadClipboard, WriteClipboard
from .files import DeleteFile, ListDir, ReadFile, WriteFile
from .inputauto import Click, MoveMouse, Screenshot, TypeText
from .net import HttpRequest
from .paths import PathScope
from .registry import ActionRegistry
from .shell import RunCommand
from .system import GetSystemState, LockScreen, Notify, SetVolume, Sleep

__all__ = ["all_actions", "build_context", "build_registry", "TIER_TABLE"]


def all_actions() -> list[Action]:
    """Every built-in action, in the order the manifest will list them."""
    return [
        # System
        GetSystemState(),
        SetVolume(),
        Notify(),
        LockScreen(),
        Sleep(),
        # Apps and windows
        LaunchApp(),
        OpenUrl(),
        ListWindows(),
        FocusWindow(),
        # Files
        ReadFile(),
        WriteFile(),
        ListDir(),
        DeleteFile(),
        # Clipboard
        ReadClipboard(),
        WriteClipboard(),
        # Network
        HttpRequest(),
        # Shell
        RunCommand(),
        # Input automation
        TypeText(),
        Click(),
        MoveMouse(),
        Screenshot(),
    ]


#: The table again, flat, as a review artefact and a test fixture. If this and
#: :func:`all_actions` ever disagree, ``tests/test_actions.py`` fails.
TIER_TABLE: dict[str, ActionTier] = {
    "get_system_state": ActionTier.AUTO,
    "set_volume": ActionTier.AUTO,
    "notify": ActionTier.AUTO,
    "lock_screen": ActionTier.CONFIRM,
    "sleep": ActionTier.CONFIRM,
    "launch_app": ActionTier.AUTO,
    "open_url": ActionTier.AUTO,
    "list_windows": ActionTier.AUTO,
    "focus_window": ActionTier.NOTIFY,
    "read_file": ActionTier.AUTO,
    "write_file": ActionTier.NOTIFY,
    "list_dir": ActionTier.AUTO,
    "delete_file": ActionTier.CONFIRM,
    "read_clipboard": ActionTier.NOTIFY,
    "write_clipboard": ActionTier.NOTIFY,
    "http_request": ActionTier.NOTIFY,
    "run_command": ActionTier.CONFIRM,
    "type_text": ActionTier.CONFIRM,
    "click": ActionTier.CONFIRM,
    "move_mouse": ActionTier.CONFIRM,
    "screenshot": ActionTier.NOTIFY,
}


def build_context(config: Config) -> ActionContext:
    scope = PathScope(config.file_roots or (config.workspace,))
    scope.ensure_roots()
    allowed_hosts = tuple(h for h in (config.server_host,) if h)
    return ActionContext(config=config, scope=scope, allowed_hosts=allowed_hosts)


def build_registry(
    config: Config,
    policy: PolicyProvider,
    audit: AuditLog,
    consent: ConsentGateway | None = None,
    ctx: ActionContext | None = None,
) -> ActionRegistry:
    """Wire the whole action layer together."""
    registry = ActionRegistry(
        ctx or build_context(config), policy=policy, audit=audit, consent=consent
    )
    registry.register_all(all_actions())
    return registry

File: android-app/app/src/main/kotlin/ai/jarvis/app/automation/actions/builtin/Builtins.kt
package ai.jarvis.app.automation.actions.builtin

import android.content.Context
import ai.jarvis.app.automation.AutomationBridge
import ai.jarvis.app.automation.actions.ActionEnv
import ai.jarvis.app.automation.actions.ActionRegistry
import ai.jarvis.app.automation.actions.ApprovalGateway
import ai.jarvis.app.automation.actions.JarvisAction
import ai.jarvis.app.automation.actions.NoPermissionGateway
import ai.jarvis.app.automation.actions.PermissionGateway
import ai.jarvis.app.automation.actions.UiApprovalGateway
import ai.jarvis.app.automation.actions.UiPermissionGateway
import ai.jarvis.app.automation.actions.asBridgeDispatcher
import ai.jarvis.app.automation.audit.AuditLog
import ai.jarvis.app.automation.policy.ActionTier
import ai.jarvis.app.automation.policy.PolicyProvider
import ai.jarvis.app.automation.policy.PolicyStore

/**
 * The local action table — the authority on what this phone can do and how
 * dangerous each of those things is. The server's `tier` field can raise these
 * numbers for a single command; nothing can lower them.
 *
 * Adding an action is a two-line change here plus a row in `docs/actions.md`.
 * Adding one at the wrong tier is a security bug, so the table is deliberately
 * in one place where it can be reviewed as a whole.
 */
object Builtins {

    /** Every built-in action, in the order the manifest will list them. */
    fun all(): List<JarvisAction> = buildList {
        // Device and system
        add(GetDeviceState)
        add(SetVolume)
        add(SetRingerMode)
        add(ToggleDnd)
        add(SetBrightness)
        add(ToggleTorch)
        add(VibrateAction)
        add(GetLocation)
        add(GetSensors)

        // Apps and intents
        add(LaunchApp)
        add(OpenUrl)
        add(ShareText)
        add(StartNavigation)
        add(DialNumber)
        add(OpenSettingsPanel)
        add(ListInstalledApps)
        add(KillApp)

        // Media
        addAll(MediaActions.all)

        // The Tasker rows closed by M61 (ParityActions.kt), one add each so the
        // table mirror can see every registration.
        add(ShowToast)
        add(SetAutoBrightness)
        add(SetRotationLock)
        add(SetScreenTimeout)
        add(GetNetworkInfo)
        add(SendIntent)
        add(LaunchShortcut)
        add(MediaControl)
        add(MediaNowPlaying)
        add(PlayMedia)
        add(SetWallpaper)
        add(RecordAudio)
        add(SetBluetooth)
        // The last six rows (CameraActions.kt, TelephonyLogActions.kt,
        // NfcActions.kt): a permission the app now asks for, or a camera.
        add(TakePhoto)
        add(ScanCode)
        add(ReadSms)
        add(ReadCallLog)
        add(EndCall)
        add(NfcRead)
        add(NfcWrite)

        // Comms
        add(SendSms)
        add(PlaceCall)
        add(ReadContacts)
        add(SendNotification)

        // Calendar and clock
        add(ReadCalendar)
        add(CreateCalendarEvent)
        add(SetAlarm)
        add(SetTimer)
        add(SetReminder)
        add(ListReminders)
        add(CancelReminder)

        // Files and clipboard
        add(ReadFile)
        add(WriteFile)
        add(ListFiles)
        add(DeleteFile)
        add(ReadClipboard)
        add(WriteClipboard)

        // Network
        add(HttpRequest)

        // Shell
        add(RunShell)

        // Delegated to the accessibility service
        addAll(UiActions.all)
    }

    /**
     * The local tier of one action id, straight from the table above, or null
     * when this build has never heard of the id.
     *
     * Static — it is the declared tier, not the per-invocation one that
     * `tierFor(params)` may raise. Used by [PolicyStore] so it can refuse to
     * store an `allow_always` for a Tier-3 action without being told the tier.
     */
    fun tierOf(actionId: String): ActionTier? = TIERS[actionId]

    private val TIERS: Map<String, ActionTier> by lazy {
        all().associate { it.id to it.tier }
    }

    /**
     * Registry wired to explicit collaborators — the form tests use.
     *
     * [permissions] defaults to [NoPermissionGateway], which knows of nothing
     * missing and can ask nobody: a unit test has no PackageManager worth
     * consulting and no Activity to raise a dialog from, so the permission step
     * stays out of the way. The app passes a real one — see [standard].
     */
    fun registry(
        context: Context,
        policy: PolicyProvider,
        audit: AuditLog,
        approvals: ApprovalGateway,
        permissions: PermissionGateway = NoPermissionGateway
    ): ActionRegistry =
        ActionRegistry(context, policy, audit, approvals, permissions).registerAll(all())

    /**
     * Registry wired the way the app runs it: SharedPreferences policy store,
     * JSONL audit log, and the full-screen consent UI.
     *
     * Build this once at startup (`AutomationRuntime.ensure` does) and hand the
     * same instance to the WebSocket client and the settings screen. Three
     * pieces of wiring happen here rather than being left to a caller who might
     * forget, because forgetting any of them is silent:
     *
     *  1. [ActionEnv.refreshFromConfig] — the jarvis-core host that
     *     `http_request` is allowed to reach, the notification-listener
     *     component the media actions prefer, the app version.
     *  2. The policy store is given the action table, so its Tier-3 guard works
     *     without the caller supplying a tier.
     *  3. [AutomationBridge.dispatcher] is filled in. `JarvisChannel` reads that
     *     slot for every `device_command`; with it empty every command from the
     *     server is answered `unsupported` and the phone does nothing at all.
     *  4. [UiPermissionGateway] is filled in. Without it every action that needs
     *     a dangerous Android permission answers `permission … not granted` and
     *     no dialog is ever shown — which is exactly what shipped, for the whole
     *     life of the app, for SMS, calls, contacts, calendar and location.
     */
    fun standard(context: Context, approvals: ApprovalGateway? = null): ActionRegistry {
        val appContext = context.applicationContext
        ActionEnv.refreshFromConfig(appContext)
        return registry(
            appContext,
            PolicyStore(appContext, ::tierOf),
            AuditLog(appContext),
            approvals ?: UiApprovalGateway(appContext),
            UiPermissionGateway(appContext)
        ).also { AutomationBridge.dispatcher = it.asBridgeDispatcher() }
    }
}
